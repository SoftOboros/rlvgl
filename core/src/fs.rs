//! Asset loading interfaces for filesystem-backed content.
//!
//! This module provides traits used by the optional `fs` feature to source
//! assets such as fonts or images from an underlying filesystem.

use alloc::boxed::Box;

/// Errors that can occur during filesystem operations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FsError {
    /// Underlying device reported an error.
    Device,
    /// Provided path was invalid.
    InvalidPath,
    /// File or directory was not found.
    NoSuchFile,
}

/// Block device abstraction used by the filesystem layer.
///
/// Implementors provide sector-based access to a storage medium.
pub trait BlockDevice {
    /// Read blocks starting at `lba` into `buf`.
    fn read_blocks(&mut self, lba: u64, buf: &mut [u8]) -> Result<(), FsError>;

    /// Write blocks starting at `lba` from `buf`.
    ///
    /// Implementations may leave this unimplemented if the device is read-only.
    fn write_blocks(&mut self, lba: u64, buf: &[u8]) -> Result<(), FsError>;

    /// Return the logical block size in bytes.
    fn block_size(&self) -> usize;

    /// Return the total number of addressable blocks.
    fn num_blocks(&self) -> u64;

    /// Flush any buffered data to the underlying device.
    fn flush(&mut self) -> Result<(), FsError>;
}

/// Error type returned by asset operations.
#[derive(Debug, Clone)]
pub enum AssetError {
    /// Underlying filesystem error.
    Fs(FsError),
}

/// Reader trait for streaming asset data.
pub trait AssetRead {
    /// Read data into `out`, returning the number of bytes read.
    fn read(&mut self, out: &mut [u8]) -> Result<usize, AssetError>;

    /// Total length of the asset in bytes.
    fn len(&self) -> usize;

    /// Return `true` if the asset has a length of zero bytes.
    fn is_empty(&self) -> bool;

    /// Seek to an absolute byte position within the asset.
    fn seek(&mut self, pos: u64) -> Result<u64, AssetError>;
}

/// Source of assets such as fonts or images.
pub trait AssetSource {
    /// Open an asset by logical path, e.g., `"fonts/regular.bin"`.
    fn open<'a>(&'a self, path: &str) -> Result<Box<dyn AssetRead + 'a>, AssetError>;

    /// Determine whether an asset at `path` exists.
    fn exists(&self, path: &str) -> bool;

    /// List the contents of `dir`, returning an iterator over asset entries.
    fn list(&self, dir: &str) -> Result<AssetIter, AssetError>;
}

/// Iterator over asset entries returned by [`AssetSource::list`].
pub struct AssetIter;

impl Iterator for AssetIter {
    type Item = (); // placeholder until fleshed out

    fn next(&mut self) -> Option<Self::Item> {
        None
    }
}

/// Manager that provides convenient typed loading helpers.
pub struct AssetManager<S: AssetSource> {
    source: S,
}

impl<S: AssetSource> AssetManager<S> {
    /// Create a new [`AssetManager`] from an [`AssetSource`].
    pub fn new(source: S) -> Self {
        Self { source }
    }

    /// Open a raw asset stream at `path`.
    pub fn open(&self, path: &str) -> Result<Box<dyn AssetRead + '_>, AssetError> {
        self.source.open(path)
    }
}
