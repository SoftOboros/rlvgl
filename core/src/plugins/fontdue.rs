//! Glyph rasterization using `fontdue`.
use alloc::vec::Vec;
pub use fontdue::Metrics;
use fontdue::{Font, FontResult, FontSettings};

/// Rasterize `ch` from the provided font data at the given pixel height.
///
/// Returns a grayscale bitmap along with its associated [`Metrics`]
/// describing placement and advance information.
///
/// The bitmap contains alpha values in row-major order which callers may use
/// to blend the glyph with an arbitrary text color.
pub fn rasterize_glyph(font_data: &[u8], ch: char, px: f32) -> FontResult<(Vec<u8>, Metrics)> {
    let font = Font::from_bytes(font_data, FontSettings::default())?;
    let (metrics, bitmap) = font.rasterize(ch, px);
    Ok((bitmap, metrics))
}

#[cfg(test)]
mod tests {
    use super::*;

    const FONT_DATA: &[u8] = include_bytes!("../../../lvgl/scripts/built_in_font/DejaVuSans.ttf");

    #[test]
    fn rasterize_a() {
        let (bitmap, metrics) = rasterize_glyph(FONT_DATA, 'A', 16.0).unwrap();
        assert_eq!(bitmap.len(), metrics.width * metrics.height);
        assert!(metrics.width > 0 && metrics.height > 0);
    }
}
