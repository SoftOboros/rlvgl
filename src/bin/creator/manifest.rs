//! Manifest structures for rlvgl-creator.
//!
//! Defines the `Manifest` schema used by CLI subcommands along with supporting
//! types for assets, groups, packages, and targets.

use std::collections::BTreeMap;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Manifest describing assets and related metadata.
#[derive(Serialize, Deserialize, JsonSchema)]
pub(crate) struct Manifest {
    #[serde(default = "default_manifest_version")]
    pub(crate) version: u8,
    #[serde(default)]
    pub(crate) packages: BTreeMap<String, Package>,
    #[serde(default)]
    pub(crate) groups: BTreeMap<String, Group>,
    #[serde(default)]
    pub(crate) features: BTreeMap<String, Vec<String>>,
    #[serde(default)]
    pub(crate) expose: BTreeMap<String, String>,
    #[serde(default)]
    pub(crate) targets: Vec<Target>,
    #[serde(default)]
    pub(crate) assets: Vec<Asset>,
}

fn default_manifest_version() -> u8 {
    1
}

impl Default for Manifest {
    fn default() -> Self {
        Self {
            version: default_manifest_version(),
            packages: BTreeMap::new(),
            groups: BTreeMap::new(),
            features: BTreeMap::new(),
            expose: BTreeMap::new(),
            targets: Vec::new(),
            assets: Vec::new(),
        }
    }
}

/// Asset entry with path, hash, and generated constant name.
#[derive(Serialize, Deserialize, JsonSchema)]
pub(crate) struct Asset {
    /// Generated constant or feature name.
    #[serde(default)]
    pub(crate) name: String,
    /// Relative path to the asset.
    pub(crate) path: String,
    /// Blake3 hash of the asset contents.
    pub(crate) hash: String,
    /// SPDX license identifier for this asset.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) license: Option<String>,
}

/// Group of assets sharing an optional license.
#[derive(Default, Serialize, Deserialize, JsonSchema)]
pub(crate) struct Group {
    /// Assets belonging to this group.
    #[serde(default)]
    pub(crate) assets: Vec<String>,
    /// SPDX license identifier applying to the group.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) license: Option<String>,
}

/// Target output configuration.
#[derive(Serialize, Deserialize, JsonSchema)]
pub(crate) struct Target {
    /// Target name.
    pub(crate) name: String,
    /// Directory where assets will be vendored.
    pub(crate) vendor_dir: String,
}

/// Feature entry emitted by the `sync` command.
#[derive(Serialize)]
pub(crate) struct FeatureEntry {
    /// Feature name.
    pub(crate) name: String,
    /// Dependent features.
    pub(crate) deps: Vec<String>,
}

/// Package grouping helper.
#[derive(Default, Serialize, Deserialize, JsonSchema)]
pub(crate) struct Package {
    /// Assets in this package.
    #[serde(default)]
    pub(crate) assets: Vec<String>,
}
