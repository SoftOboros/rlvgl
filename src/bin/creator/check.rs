//! Check command for rlvgl-creator.
//!
//! Validates manifest entries against asset files and naming policies.

use std::collections::HashSet;
use std::fs;
use std::path::Path;

use anyhow::{Result, bail};
use blake3::Hasher;

use crate::manifest::Manifest;
use crate::scan;
use crate::util::{const_name, valid_root};
use serde_yaml;

/// Verify manifest paths, hashes, and names, optionally fixing issues.
pub(crate) fn run(root: &Path, manifest_path: &Path, fix: bool) -> Result<()> {
    if fix {
        scan::run(root, manifest_path)?;
    }

    if !manifest_path.exists() {
        bail!("`{}` not found", manifest_path.display());
    }

    let contents = fs::read_to_string(manifest_path)?;
    let mut manifest: Manifest = serde_yaml::from_str(&contents)?;

    let mut issues = Vec::new();
    let mut changed = false;
    let mut seen_paths = HashSet::new();
    let mut seen_names = HashSet::new();
    let mut new_assets = Vec::new();

    for mut asset in manifest.assets.into_iter() {
        if !seen_paths.insert(asset.path.clone()) {
            if fix {
                println!("Removed duplicate entry `{}`", asset.path);
                changed = true;
                continue;
            }
            issues.push(format!("Duplicate entry `{}`", asset.path));
        }
        if !seen_names.insert(asset.name.clone()) {
            if fix {
                println!("Removed duplicate name `{}`", asset.name);
                changed = true;
                continue;
            }
            issues.push(format!("Duplicate name `{}`", asset.name));
        }

        let asset_path = root.join(&asset.path);
        if !asset_path.exists() {
            if fix {
                println!("Removed missing file `{}`", asset.path);
                changed = true;
                continue;
            }
            issues.push(format!("Missing file `{}`", asset.path));
        }

        if !valid_root(&asset.path) {
            if fix {
                println!("Removed invalid root `{}`", asset.path);
                changed = true;
                continue;
            }
            issues.push(format!("Invalid root `{}`", asset.path));
        }

        let data = fs::read(&asset_path)?;
        let mut hasher = Hasher::new();
        hasher.update(&data);
        let hash = hasher.finalize().to_hex().to_string();
        if hash != asset.hash {
            if fix {
                println!("Updated hash for `{}`", asset.path);
                asset.hash = hash;
                changed = true;
            } else {
                issues.push(format!("Hash mismatch for `{}`", asset.path));
            }
        }

        let expected_name = const_name(&asset.path);
        if asset.name != expected_name {
            if fix {
                println!("Updated name for `{}`", asset.path);
                asset.name = expected_name;
                changed = true;
            } else {
                issues.push(format!(
                    "Name mismatch for `{}`: expected `{}`, found `{}`",
                    asset.path, expected_name, asset.name
                ));
            }
        }

        new_assets.push(asset);
    }

    manifest.assets = new_assets;

    if fix && changed {
        fs::write(manifest_path, serde_yaml::to_string(&manifest)?)?;
    }

    if issues.is_empty() {
        println!("All assets valid");
        Ok(())
    } else {
        for issue in issues {
            eprintln!("Error: {}", issue);
        }
        bail!("Asset check failed");
    }
}
