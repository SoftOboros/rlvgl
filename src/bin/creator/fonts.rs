//! Fonts packing utilities for rlvgl-creator.
//!
//! Packs TTF/OTF fonts into bitmap `.bin` files and matching metrics `.json` descriptors.

use std::fs;
use std::path::Path;

use anyhow::{Result, anyhow};
use fontdue::{Font, FontSettings};
use serde::Serialize;
use walkdir::WalkDir;

use crate::check;

#[derive(Serialize)]
struct GlyphMetric {
    ch: char,
    width: usize,
    height: usize,
    advance: f32,
    offset: usize,
}

/// Pack font files under `root` into binary and JSON outputs.
pub(crate) fn pack(root: &Path, manifest_path: &Path, size: f32, chars: &str) -> Result<()> {
    for entry in WalkDir::new(root).into_iter().filter_map(|e| e.ok()) {
        let path = entry.path();
        if !entry.file_type().is_file() {
            continue;
        }
        match path.extension().and_then(|e| e.to_str()) {
            Some("ttf") | Some("otf") => {}
            _ => continue,
        }

        let data = fs::read(path)?;
        let font = Font::from_bytes(data, FontSettings::default()).map_err(|e| anyhow!(e))?;

        let mut bin = Vec::new();
        let mut metrics = Vec::new();
        let mut offset = 0usize;

        for ch in chars.chars() {
            let (m, bitmap) = font.rasterize(ch, size);
            metrics.push(GlyphMetric {
                ch,
                width: m.width,
                height: m.height,
                advance: m.advance_width,
                offset,
            });
            bin.extend_from_slice(&bitmap);
            offset += bitmap.len();
        }

        let stem = path.file_stem().unwrap().to_string_lossy();
        let bin_path = path.with_file_name(format!("{}-{}.bin", stem, size as usize));
        let json_path = path.with_file_name(format!("{}-{}.json", stem, size as usize));
        fs::write(&bin_path, &bin)?;
        fs::write(&json_path, serde_json::to_vec(&metrics)?)?;
        println!(
            "Packed {} -> {}, {}",
            path.display(),
            bin_path.display(),
            json_path.display()
        );
    }

    // Refresh manifest with new font assets.
    check::run(root, manifest_path, true)?;
    Ok(())
}
#[cfg(all(test, feature = "regression"))]
mod tests {
    use super::*;
    use crate::manifest::{Group, Manifest};
    use blake3::hash;
    use tempfile::tempdir;

    #[test]
    fn pack_generates_stable_bin_and_json() {
        let dir = tempdir().unwrap();
        let root = dir.path();
        let fonts_dir = root.join("fonts");
        fs::create_dir_all(&fonts_dir).unwrap();
        fs::copy(
            Path::new("lvgl/scripts/built_in_font/Montserrat-Medium.ttf"),
            fonts_dir.join("sample.ttf"),
        )
        .unwrap();

        let mut manifest = Manifest::default();
        manifest.groups.insert(
            "sample".into(),
            Group {
                assets: vec![
                    "fonts/sample.ttf".into(),
                    "fonts/sample-12.bin".into(),
                    "fonts/sample-12.json".into(),
                ],
                license: Some("MIT".into()),
            },
        );
        let manifest_path = root.join("manifest.yml");
        fs::write(&manifest_path, serde_yaml::to_string(&manifest).unwrap()).unwrap();

        pack(root, &manifest_path, 12.0, "AB").unwrap();

        let bin_hash = hash(&fs::read(fonts_dir.join("sample-12.bin")).unwrap()).to_hex();
        let json_hash = hash(&fs::read(fonts_dir.join("sample-12.json")).unwrap()).to_hex();
        insta::assert_snapshot!("font_hashes", format!("{}\n{}", bin_hash, json_hash));
    }
}
