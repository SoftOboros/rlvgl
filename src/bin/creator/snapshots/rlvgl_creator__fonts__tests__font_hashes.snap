---
source: src/bin/creator/fonts.rs
expression: "format!(\"{}\\n{}\", bin_hash, json_hash)"
---
652cac0891936be91df5f7775a567234f7f543c6f54d1f232d72f0128b5de3ed
8fe460aba6c13ade3e7ca8e7ad674e294c71228e9e0477991e67074cea169792
